File: 64minimumPathSum/minimumPathSum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        *dp, = itertools.accumulate(grid[0])
        for row in itertools.islice(grid, 1, None):
            dp[0] += row[0]
            for column in range(1, len(row)): dp[column] = min(itertools.islice(dp, column - 1, column + 1)) + row[column]
        return dp[-1]

File: 64minimumPathSum/minimumPathSum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int[] dp = grid[0];
        Arrays.parallelPrefix(dp, (a,b) -> a + b);
        Stream.of(grid).skip(1).forEach(row -> {
            dp[0] += row[0];
            IntStream.range(1, row.length).forEach(column -> dp[column] = IntStream.range(column - 1, column + 1).map($ -> dp[$]).min().orElse(0) + row[column]);});
        return dp[dp.length - 1];
    }
}
